### DISCUSSION TITLE
Guard In Code Still Passed

### DISCUSSION THREAD

**Discussion by pr0ph0z** [prisma/prisma#15573] [Q&A]
Hi there, I have a function that will check the input, read the data from the database, and decide whether it's passed the check or not.

```js
async function startTest({ test_id, group_id, student_id }) {
  const findStartedTest = await prisma.runningTest.findFirst({
    where: {
      testId: test_id,
      groupId: group_id,
      studentId: student_id,
      deletedAt: null
    }
  });

  if (findStartedTest !== null) {
    return {
      ...requestResponse.unprocessable_entity,
      message: "You've started this test!"
    };
  }

  const currentTimestamp = new Date();
  const runningTestId = v4();

  await prisma.runningTest.create({
    data: {
      id: runningTestId,
      test: {
        connect: {
          id: test_id
        }
      },
      group: {
        connect: {
          id: group_id
        }
      },
      student: {
        connect: {
          id: student_id
        }
      },
      testStart: currentTimestamp,
      testEnd: addMinutes(currentTimestamp, test.duration)
    }
  });

  return {
    ...requestResponse.success,
    data: {
      id: runningTestId,
      test_end: addMinutes(currentTimestamp, test.duration)
    }
  };
}
```
The thing is, I found some data that created at the same time (the difference are only few milliseconds)

![image](https://user-images.githubusercontent.com/17337978/192714124-7596c368-ccff-442e-ae92-e5506a151834.png)

I tried to reproduce it with `Promise.allSettled`

```js
const tests = await Promise.allSettled([
  testService.startTest({
    test_id: "8afd180c-8605-47a4-a44d-f021b5642e0a",
    group_id: "fa4097bc-6612-4843-9d3e-62cd906c122c",
    student_id: "4e6c428e-d1df-4907-bf12-bea6e872ae34"
  }),
  testService.startTest({
    test_id: "8afd180c-8605-47a4-a44d-f021b5642e0a",
    group_id: "fa4097bc-6612-4843-9d3e-62cd906c122c",
    student_id: "4e6c428e-d1df-4907-bf12-bea6e872ae34"
  }),
  testService.startTest({
    test_id: "8afd180c-8605-47a4-a44d-f021b5642e0a",
    group_id: "fa4097bc-6612-4843-9d3e-62cd906c122c",
    student_id: "4e6c428e-d1df-4907-bf12-bea6e872ae34"
  })
]);
console.log(tests);
```

And they are all fulfilled

![image](https://user-images.githubusercontent.com/17337978/192714933-7a4aad1e-201a-487e-8b0e-93989a8ed373.png)

Maybe it's not a Prisma-related problem but is anyone has a workaround to fix the problem?

**Comment by pr0ph0z** [comment]
One thing that comes into my mind is table locking and I think it's related to this table locking issue https://github.com/prisma/prisma/issues/3088

**Comment by austincrim** [comment]
Hey @pr0ph0z 👋,

You should wrap the logic in the function inside an [interactive transaction](https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide#interactive-transactions-in-preview).

**Reply by pr0ph0z** [reply]
Hi @austincrim, thank you for replying. But unfortunately, it's still not working in the interactive transaction. Or did I miss something?

Here's my code:

```js
const { PrismaClient } = require("@prisma/client");
const prisma = new PrismaClient();
const { v4 } = require("uuid");
const addMinutes = require("date-fns/addMinutes");

async function startTest({ test_id, group_id, student_id }) {
  return await prisma.$transaction(async (tx) => {
    const findStartedTest = await tx.runningTest.findFirst({
      where: {
        testId: test_id,
        groupId: group_id,
        studentId: student_id,
        deletedAt: null,
      },
    });

    if (findStartedTest !== null) {
      throw new Error("You've started this test!");
    }

    const currentTimestamp = new Date();
    const runningTestId = v4();

    await tx.runningTest.create({
      data: {
        id: runningTestId,
        test: {
          connect: {
            id: test_id,
          },
        },
        group: {
          connect: {
            id: group_id,
          },
        },
        student: {
          connect: {
            id: student_id,
          },
        },
        testStart: currentTimestamp,
        testEnd: addMinutes(currentTimestamp, test.duration),
      },
    });

    return {
      data: {
        id: runningTestId,
        test_end: addMinutes(currentTimestamp, test.duration),
      },
    };
  });
}

(async () => {
  const tests = await Promise.allSettled([
    startTest({
      test_id: "8afd180c-8605-47a4-a44d-f021b5642e0a",
      group_id: "fa4097bc-6612-4843-9d3e-62cd906c122c",
      student_id: "4e6c428e-d1df-4907-bf12-bea6e872ae34",
    }),
    startTest({
      test_id: "8afd180c-8605-47a4-a44d-f021b5642e0a",
      group_id: "fa4097bc-6612-4843-9d3e-62cd906c122c",
      student_id: "4e6c428e-d1df-4907-bf12-bea6e872ae34",
    }),
    startTest({
      test_id: "8afd180c-8605-47a4-a44d-f021b5642e0a",
      group_id: "fa4097bc-6612-4843-9d3e-62cd906c122c",
      student_id: "4e6c428e-d1df-4907-bf12-bea6e872ae34",
    }),
  ]);
  console.log(tests);
})();
```

**Reply by austincrim** [reply]
Hm, that looks right to me. The issue you linked does seem relevant, would you mind adding a comment to it with your specific use case?

In the meantime, you might have to add some extra logic in your app code to guard against this scenario.

**Reply by vladi-stevanovic** [reply]
👋 Hello @pr0ph0z - I wanted to follow up to see if you were able to resolve this by adding extra logic in your app?
If this is still a pain point I recommend adding a comment with your use case to issue [#3088 ](https://github.com/prisma/prisma/issues/3088)!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
